### ISSUE TITLE
`colors="true"` in XML configuration file does not unconditionally enable colored output

### ISSUE THREAD

**Issue by Bilge** [sebastianbergmann/phpunit#5881]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 10.5.24
| PHP version         | 8.2.15
| Installation Method | Composer

#### Summary

PHPUnit 9 `--coverage-text` with `colors="true"` would output the report in colour, making it easy to identify which files were uncovered. PHPUnit 10 does not do this.

#### Current behavior

Colours are gone 😞 

![image](https://github.com/sebastianbergmann/phpunit/assets/470626/e10953b5-978e-46c3-a747-945fd2049a17)

#### How to reproduce

`phpunit --coverage-text`

#### Expected behavior

Colours.

![image](https://github.com/sebastianbergmann/phpunit/assets/470626/0299b7c4-3ab7-4077-bdb8-5aacb9fcdf54)

**Comment by sebastianbergmann** [comment]
I cannot reproduce that with the test suite of https://github.com/sebastianbergmann/raytracer, for example: 
![grafik](https://github.com/sebastianbergmann/phpunit/assets/25218/0cc149ec-4805-4fca-a939-fffee97174f3)

**Comment by sebastianbergmann** [comment]
Thank you for your report.

Please provide a **minimal**, **self-contained**, reproducing test case that shows the problem you are reporting.

Without such a minimal, self-contained, reproducing test case I will not be able to investigate this issue.

**Comment by Bilge** [comment]
I think it is less about the test case and more about the environment. I was able to get it to work in WSL 2 (Ubuntu), but it does not work in GitHub Actions and it does not work in [Git Bash](https://git-scm.com/download/win) (mintty) on Windows.

**Comment by sebastianbergmann** [comment]
I am sorry to say this, but this report is not actionable for me. I cannot and will not work on this myself, as I do not know how. I somebody, maybe you, finds the cause for this issue in their environment and sends a pull request then I will, of course, consider that.

**Comment by Bilge** [comment]
Cannot = false. Will not = noted. You have access to GitHub Actions, same as me. Nevertheless, I'll try to get to the bottom of it myself. Alone.

**Comment by sebastianbergmann** [comment]
Terminal capabilities are checked by `sebastian/environment`. PHPUnit 9 uses `^5` of that library, PHPUnit 10 uses `^6`.

Looking at https://github.com/sebastianbergmann/environment/compare/5.1.5...6.1.0#files_bucket, I do not see any code changes between the version used by PHPUnit 9 and PHPUnit 10 that are related to detecting terminal capabilities. https://github.com/sebastianbergmann/environment/blob/5.1.5/src/Console.php#L55-L77 and https://github.com/sebastianbergmann/environment/blob/6.1.0/src/Console.php#L55-L77 are identical, AFAICS.

**Comment by Bilge** [comment]
Setting `TERM=xterm` fixes it for mintty (it was previously xterm-256color). Still no idea what changed to break GitHub Actions.

**Comment by sebastianbergmann** [comment]
Looks like the original code in Symfony Console that I copied/adapted has changed. [It now looks like so.](https://github.com/symfony/symfony/blob/v7.1.2/src/Symfony/Component/Console/Output/StreamOutput.php#L90-L120)

However, that does not explain the difference between PHPUnit 9 and PHPUnit 10 in your environment.

**Comment by Bilge** [comment]
At some point (somewhat recently) I manually changed the TERM setting from `xterm` to `xterm-256color`, so that is resolved. It is just GitHub Actions that remains unresolved.

**Comment by Bilge** [comment]
I think a valid question would be, _why does PHPUnit even attempt to do capabilities detection when `colors=true`?_ If we're explicitly enabling it, the software should take that as read. I would expect the current behaviour only when `colors=auto`. Compare to `ls`:

       --color[=WHEN]
              colorize the output; WHEN can be 'always' (default if omitted), 'auto', or 'never'; more info below

Further, the whole notion of capabilities detection seems somewhat of a nonsense because all the escapes are hardcoded anyway; it's not like we're pulling the correct escapes from a term capabilities database (as should be done if we're being strictly correct).

GitHub Actions environment is a tricky case because it does not behave like a valid terminal environment supporting colour, yet it does support it anyway. Correct colour capabilities detection (e.g. using `tput`) generally fails there. Of course, it would still be nice to get to the bottom of what changed between 9 and 10.

**Comment by Bilge** [comment]
Although I haven't been able to step debug/trace it, it seems clear from this behaviour that in PHPUnit 9, whether due to a bug or otherwise, capabilities simply weren't being checked. In PHPUnit 10, whether internally or otherwise, capabilities are now being consulted before outputting colours, as we can see by colours now being disabled when the (limited) environment detection kicks in for unsupported environments. Therefore, looking for changes in the code for capabilities support is a red herring, because in the previous version, capabilities support wasn't even being consulted at all.

I still think the previous behaviour was correct. If we want to do environment detection, a new `colors=auto` mode would be preferable (and update the old detection code copied from Symfony). This is backwards-compatible behaviour then.

**Comment by Bilge** [comment]
To get colour support back in all environments, (i.e. the old PHPUnit 9 behaviour) was must set `TERM_PROGRAM=Hyper` in the environment. e.g.

```json
    "scripts": {
         "test": ["@putenv TERM_PROGRAM=Hyper", "phpunit"],
    }
```

**Comment by Bilge** [comment]
> Looks like the original code in Symfony Console that I copied/adapted has changed. [It now looks like so.](https://github.com/symfony/symfony/blob/v7.1.2/src/Symfony/Component/Console/Output/StreamOutput.php#L90-L120)

What are we going to do about this? Just copy/paste the updated code, or try to convince them to make this code public?

**Comment by llaville** [comment]
@Bilge Thanks for your [solution](https://github.com/sebastianbergmann/phpunit/issues/5881#issuecomment-2211781831), it helps me to re-introduced PHPUnit code coverage feature that I've not used since a long time.

And it works well on GitHub Actions and WSL 2 (for my locally tests) on PHPUnit 10.5 !

TL;DR but more details (with screenshots) are given at https://github.com/llaville/box-manifest/discussions/12

**Comment by sebastianbergmann** [comment]
I investigated this and found two separate issues contributing to the problem.

### 1. `colors="true"` in XML maps to `auto` instead of `always`

In `src/TextUI/Configuration/Xml/Loader.php`, `colors="true"` is mapped to `COLOR_AUTO`, not `COLOR_ALWAYS`. This means that even when a user explicitly enables colors in their configuration, PHPUnit still checks terminal capability detection via `Console::hasColorSupport()` and may decide not to use colors.

The fix is to map `colors="true"` to `COLOR_ALWAYS`. When a user explicitly enables colors, that should be respected unconditionally.

### 2. `sebastian/environment` `hasColorSupport()` is stricter than it needs to be

The `hasColorSupport()` method in `sebastian/environment` was adapted from Symfony's `StreamOutput`, but the Symfony version has since been updated to support the `FORCE_COLOR` environment variable (see https://force-color.org/). The current `sebastian/environment` code does not support `FORCE_COLOR`, which means there is no way for users in non-TTY environments (like GitHub Actions) to force color output even when their environment supports it.

I have opened an issue on `sebastian/environment` to add `FORCE_COLOR` support: https://github.com/sebastianbergmann/environment/issues/80

### Summary

- **PHPUnit fix**: `colors="true"` in `phpunit.xml` now maps to `always` (unconditional color output) instead of `auto` (capability-detected). This matches user intent: explicitly enabling colors should not be second-guessed by capability detection.
- **sebastian/environment fix**: Add support for the `FORCE_COLOR` environment variable, following the https://force-color.org/ convention that Symfony and many other tools already support. This improves the `auto` detection for CI environments like GitHub Actions.

**Comment by Bilge** [comment]
Only took two years to become humble enough to acknowledge fault.
